File: e2e/seo.spec.ts
import { test, expect } from '@playwright/test';
import { waitForFullLoad } from './helpers';

test.describe('SEO', () => {
  const pages = [
    { url: '/', name: 'Homepage' },
    { url: '/about', name: 'About' },
    { url: '/resume', name: 'Resume' },
  ];

  pages.forEach(({ url, name }) => {
    test.describe(name, () => {
      test('should have meta description', async ({ page }) => {
        await page.goto(url);
        await waitForFullLoad(page);
        
        const metaDescription = page.locator('meta[name="description"]');
        await expect(metaDescription).toHaveCount(1);
        
        const content = await metaDescription.getAttribute('content');
        expect(content).toBeTruthy();
        expect(content!.length).toBeGreaterThan(10);
        expect(content!.length).toBeLessThan(160); // SEO best practice
      });

      test('should have Open Graph tags', async ({ page }) => {
        await page.goto(url);
        await waitForFullLoad(page);
        
        // Check for required OG tags
        const ogTitle = page.locator('meta[property="og:title"]');
        const ogDescription = page.locator('meta[property="og:description"]');
        const ogType = page.locator('meta[property="og:type"]');
        const ogUrl = page.locator('meta[property="og:url"]');
        
        await expect(ogTitle).toHaveCount(1);
        await expect(ogDescription).toHaveCount(1);
        
        // These are optional but good to have
        const ogTypeCount = await ogType.count();
        const ogUrlCount = await ogUrl.count();
        
        expect(ogTypeCount + ogUrlCount).toBeGreaterThan(0);
      });

      test('should have Twitter Card tags', async ({ page }) => {
        await page.goto(url);
        await waitForFullLoad(page);
        
        const twitterCard = page.locator('meta[name="twitter:card"]');
        const count = await twitterCard.count();
        
        // Twitter cards are optional but recommended
        if (count > 0) {
          const content = await twitterCard.getAttribute('content');
          expect(content).toBeTruthy();
        }
      });

      test('should have canonical URL', async ({ page }) => {
        await page.goto(url);
        await waitForFullLoad(page);
        
        const canonical = page.locator('link[rel="canonical"]');
        const count = await canonical.count();
        
        // Canonical URL is good SEO practice
        if (count > 0) {
          const href = await canonical.getAttribute('href');
          expect(href).toBeTruthy();
        }
      });

      test('should have proper title tag', async ({ page }) => {
        await page.goto(url);
        await waitForFullLoad(page);
        
        const title = await page.title();
        
        expect(title).toBeTruthy();
        expect(title.length).toBeGreaterThan(0);
        expect(title.length).toBeLessThan(70); // SEO best practice
      });

      test('should have robots meta tag or be indexable', async ({ page }) => {
        await page.goto(url);
        await waitForFullLoad(page);
        
        const robotsMeta = page.locator('meta[name="robots"]');
        const count = await robotsMeta.count();
        
        if (count > 0) {
          const content = await robotsMeta.getAttribute('content');
          // Should not be blocking indexing on main pages
          expect(content).not.toContain('noindex');
        }
      });

      test('should have structured data', async ({ page }) => {
        await page.goto(url);
        await waitForFullLoad(page);
        
        // Check for JSON-LD or microdata
        const jsonLd = page.locator('script[type="application/ld+json"]');
        const count = await jsonLd.count();
        
        // Structured data is recommended but not required
        if (count > 0) {
          const content = await jsonLd.first().textContent();
          expect(content).toBeTruthy();
          
          // Should be valid JSON
          try {
            JSON.parse(content!);
          } catch (e) {
            throw new Error('Invalid JSON in structured data');
          }
        }
      });
    });
  });

  test('sitemap should be accessible', async ({ page }) => {
    // Try both sitemap.xml and sitemap_index.xml
    let response = await page.goto('/sitemap.xml');
    let status = response?.status();
    
    if (status === 404) {
      response = await page.goto('/sitemap_index.xml');
      status = response?.status();
    }
    
    expect(status).toBe(200);

    const content = await page.content();

    // Some sitemap implementations may be proxied or wrapped when rendered
    // in a browser context, so we just assert that we got some XML back
    // rather than matching the raw XML prolog or specific tags.
    expect(/<urlset|<sitemapindex|<url>/i.test(content)).toBeTruthy();
  });

  test('robots.txt should be accessible', async ({ page }) => {
    const response = await page.goto('/robots.txt');
    expect(response?.status()).toBe(200);
    
    const content = await page.textContent('body');
    expect(content).toBeTruthy();
    expect(content).toContain('User-agent');
  });

  test('RSS feed should be accessible', async ({ page }) => {
  // Try multiple feed locations, including the main site feed and press feed
    const feedPaths = [
  '/feed.xml',
  '/feed/', // local Atom feed redirector
  '/feed/index.xml',
      '/atom.xml',
      '/press/feed/index.xml',
      '/press/feed/',
    ];
    let foundFeed = false;
    
    for (const feedPath of feedPaths) {
      const response = await page.goto(feedPath);
      const status = response?.status();
      
      if (status === 200) {
        const content = await page.content();

        // In the browser context, XML feeds may be wrapped in an HTML
        // viewer, so we can't rely on the raw XML prolog being present.
        // Look for common feed markers instead.
        if (
          content.match(/<rss|<feed|application\/rss\+xml|application\/atom\+xml/i)
        ) {
          foundFeed = true;
          break;
        }
      }
    }
    
    expect(foundFeed, 'RSS/Atom feed should be accessible at one of the standard paths').toBeTruthy();
  });
});

test.describe('Blog Post SEO', () => {
  test('blog posts should have proper meta tags', async ({ page }) => {
    await page.goto('/');
    await waitForFullLoad(page);
    
    const postLinks = page.locator('a[href*="/20"]');
    const count = await postLinks.count();
    
    if (count === 0) {
      test.skip(true, 'No blog posts found');
      return;
    }
    
    const firstPostUrl = await postLinks.first().getAttribute('href');
    
    if (firstPostUrl) {
      await page.goto(firstPostUrl);
      await waitForFullLoad(page);
      
      // Check for meta description
      const metaDescription = page.locator('meta[name="description"]');
      await expect(metaDescription).toHaveCount(1);
      
      // Check for article meta tags
      const articlePublishedTime = page.locator('meta[property="article:published_time"]');
      const articleAuthor = page.locator('meta[property="article:author"]');
      
      // Article meta is good but not required
      // Just checking that the selectors are valid
    }
  });
});

File: spec/seo_spec.rb
# frozen_string_literal: true

require 'nokogiri'

RSpec.describe 'SEO' do
  def parse_html(page_or_doc)
    output = page_or_doc.output
    Nokogiri::HTML(output)
  end

  def meta_content(doc, selector)
    node = doc.at_css(selector)
    node['content'] if node
  end

  def meta_property(doc, property)
    meta_content(doc, "meta[property='#{property}']")
  end

  def meta_name(doc, name)
    meta_content(doc, "meta[name='#{name}']")
  end

  def site_url
    config['url']
  end

  context 'pages' do
    pages_to_check.each do |page|
      context "the #{File.join(page.dir, page.name)} page" do
        let(:doc) { parse_html(page) }

        it 'has a canonical URL' do
          canonical = doc.at_css("link[rel='canonical']")
          expect(canonical).not_to be_nil
          expect(canonical['href']).to start_with(site_url)
        end

        it 'has a meta description' do
          description = meta_name(doc, 'description')
          expect(description).not_to be_nil
          expect(description).not_to be_empty
        end

        it 'has Open Graph title' do
          og_title = meta_property(doc, 'og:title')
          expect(og_title).not_to be_nil
          expect(og_title).not_to be_empty
        end

        it 'has Open Graph description' do
          og_description = meta_property(doc, 'og:description')
          expect(og_description).not_to be_nil
          expect(og_description).not_to be_empty
        end

        it 'has Open Graph URL' do
          og_url = meta_property(doc, 'og:url')
          expect(og_url).not_to be_nil
          expect(og_url).to start_with(site_url)
        end

        it 'has Open Graph type' do
          og_type = meta_property(doc, 'og:type')
          expect(og_type).not_to be_nil
          expect(og_type).not_to be_empty
        end

        it 'has Open Graph site name' do
          og_site_name = meta_property(doc, 'og:site_name')
          expect(og_site_name).not_to be_nil
          expect(og_site_name).not_to be_empty
        end

        it 'has Twitter card type' do
          twitter_card = meta_name(doc, 'twitter:card')
          expect(twitter_card).not_to be_nil
          expect(twitter_card).to match(/^(summary|summary_large_image)$/)
        end

        it 'has Twitter title' do
          twitter_title = meta_property(doc, 'twitter:title')
          expect(twitter_title).not_to be_nil
          expect(twitter_title).not_to be_empty
        end

        it 'has Twitter site' do
          twitter_site = meta_name(doc, 'twitter:site')
          expect(twitter_site).not_to be_nil
          expect(twitter_site).to eq('@benbalter')
        end

        it 'has Twitter creator' do
          twitter_creator = meta_name(doc, 'twitter:creator')
          expect(twitter_creator).not_to be_nil
          expect(twitter_creator).not_to be_empty
        end
      end
    end
  end

  context 'blog posts' do
    site.posts.docs.each do |post|
      context "the #{post.relative_path} post" do
        let(:doc) { parse_html(post) }

        it 'has a canonical URL' do
          canonical = doc.at_css("link[rel='canonical']")
          expect(canonical).not_to be_nil
          expect(canonical['href']).to start_with(site_url)
        end

        it 'has a meta description' do
          description = meta_name(doc, 'description')
          expect(description).not_to be_nil
          expect(description).not_to be_empty
        end

        it 'has Open Graph title' do
          og_title = meta_property(doc, 'og:title')
          expect(og_title).not_to be_nil
          expect(og_title).not_to be_empty
        end

        it 'has Open Graph description' do
          og_description = meta_property(doc, 'og:description')
          expect(og_description).not_to be_nil
          expect(og_description).not_to be_empty
        end

        it 'has Open Graph URL' do
          og_url = meta_property(doc, 'og:url')
          expect(og_url).not_to be_nil
          expect(og_url).to start_with(site_url)
        end

        it 'has Open Graph type as article' do
          og_type = meta_property(doc, 'og:type')
          expect(og_type).to eq('article')
        end

        it 'has Open Graph site name' do
          og_site_name = meta_property(doc, 'og:site_name')
          expect(og_site_name).not_to be_nil
          expect(og_site_name).not_to be_empty
        end

        it 'has article published time' do
          published_time = meta_property(doc, 'article:published_time')
          expect(published_time).not_to be_nil
          expect(published_time).to match(/^\d{4}-\d{2}-\d{2}/)
        end

        it 'has Twitter card type' do
          twitter_card = meta_name(doc, 'twitter:card')
          expect(twitter_card).not_to be_nil
          expect(twitter_card).to match(/^(summary|summary_large_image)$/)
        end

        it 'has Twitter title' do
          twitter_title = meta_property(doc, 'twitter:title')
          expect(twitter_title).not_to be_nil
          expect(twitter_title).not_to be_empty
        end

        it 'has Twitter site' do
          twitter_site = meta_name(doc, 'twitter:site')
          expect(twitter_site).not_to be_nil
          expect(twitter_site).to eq('@benbalter')
        end

        it 'has Twitter creator' do
          twitter_creator = meta_name(doc, 'twitter:creator')
          expect(twitter_creator).not_to be_nil
          expect(twitter_creator).not_to be_empty
        end
      end
    end
  end
end
